src/Io/Transaction.php

```diff
@@ -255,25 +256,38 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
     /**
      * @param RequestInterface $request
      * @param UriInterface $location
+     * @param int $statusCode
      * @return RequestInterface
      */
-    private function makeRedirectRequest(RequestInterface $request, UriInterface $location)
+    private function makeRedirectRequest(RequestInterface $request, UriInterface $location, $statusCode)
     {
         $originalHost = $request->getUri()->getHost();
-        $request = $request
-            ->withoutHeader('Host')
-            ->withoutHeader('Content-Type')
-            ->withoutHeader('Content-Length');
+        $request = $request->withoutHeader('Host');
+
+        if ($statusCode !== 307 || $statusCode !== 308) {
+            $request = $request
+                ->withoutHeader('Content-Type')
+                ->withoutHeader('Content-Length');
+        }
 
         // Remove authorization if changing hostnames (but not if just changing ports or protocols).
         if ($location->getHost() !== $originalHost) {
             $request = $request->withoutHeader('Authorization');
         }
 
-        // naïve approach..
-        $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
+        $body = null;
+        if ($statusCode === 307 || $statusCode === 308) {
+            $method = $request->getMethod();
+            if ($request->getBody() instanceof ReadableStreamInterface) {
+                throw new RuntimeException('Unable to redirect request with streaming body');
+            }
+            $body = $request->getBody();
+        } else {
+            // naïve approach..
+            $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
+        }
 
-        return new Request($method, $location, $request->getHeaders());
+        return new Request($method, $location, $request->getHeaders(), $body);
```

Comment:
Might as well return the request returned from the withers here instead of explicitly creating a new instance, WDYT?

Reply:
```
        if (!$this->isTemporaryOrPermanentRedirect($statusCode)) {
            // naïve approach..
            $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
            $request = $request->withMethod($method)->withBody(new EmptyBodyStream());
        } elseif ($request->getBody() instanceof ReadableStreamInterface) {
            throw new \RuntimeException('Unable to redirect request with streaming body');
        }
```

This could now also be refactored to this so we don't have the double if with only the exception but I'm not sure if it's really better.